File: examples/python/vaultline.py
"""
VaultLine API Client

Usage:
    client = VaultLine('vl_your_api_key')
    rates = client.get_rates(['BTC/USD', 'ETH/USD'])
"""

import requests
from dataclasses import dataclass
from typing import Optional


@dataclass
class RateLimit:
    limit: int
    remaining: int
    reset: int


class VaultLineError(Exception):
    """API error with status code and message"""
    def __init__(self, status_code: int, message: str):
        self.status_code = status_code
        self.message = message
        super().__init__(f"[{status_code}] {message}")


class VaultLine:
    def __init__(self, api_key: str, base_url: str = "https://api.vaultline.io/api/v1"):
        self.api_key = api_key
        self.base_url = base_url
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {api_key}"
        })
        self.rate_limit: Optional[RateLimit] = None

    def _request(self, endpoint: str, params: dict = None) -> dict:
        """Make an API request and return JSON response"""
        response = self.session.get(f"{self.base_url}{endpoint}", params=params)

        # Extract rate limit info
        self.rate_limit = RateLimit(
            limit=int(response.headers.get("X-RateLimit-Limit", 0)),
            remaining=int(response.headers.get("X-RateLimit-Remaining", 0)),
            reset=int(response.headers.get("X-RateLimit-Reset", 0)),
        )

        if not response.ok:
            error = response.json()
            raise VaultLineError(
                response.status_code,
                error.get("message") or error.get("error") or f"HTTP {response.status_code}"
            )

        return response.json()

    def get_assets(self, asset_class: str = None) -> dict:
        """
        Get available symbols grouped by asset class

        Args:
            asset_class: Filter by 'fiat', 'crypto', 'stocks', or 'metals'

        Returns:
            dict with asset_classes and total_symbols
        """
        params = {}
        if asset_class:
            params["asset_class"] = asset_class
        return self._request("/assets", params)

    def get_rates(self, symbols: list[str], date: str = None) -> dict:
        """
        Get current rates for symbols

        Args:
            symbols: List of symbols, e.g. ['BTC/USD', 'AAPL']
            date: Optional ISO date (YYYY-MM-DD)

        Returns:
            dict with data array of rates
        """
        params = {"symbols": ",".join(symbols)}
        if date:
            params["date"] = date
        return self._request("/rates", params)

    def get_history(self, symbol: str, from_date: str, to_date: str) -> dict:
        """
        Get historical rates for a symbol

        Args:
            symbol: Symbol, e.g. 'BTC/USD'
            from_date: Start date (YYYY-MM-DD)
            to_date: End date (YYYY-MM-DD)

        Returns:
            dict with symbol info and history array
        """
        return self._request("/rates/history", {
            "symbol": symbol,
            "from": from_date,
            "to": to_date,
        })


# Example usage
if __name__ == "__main__":
    from datetime import datetime, timedelta

    client = VaultLine("vl_your_api_key_here")

    # Get all available assets
    assets = client.get_assets()
    print(f"Available assets: {assets['total_symbols']}")

    # Get crypto assets only
    crypto = client.get_assets("crypto")
    print(f"Crypto symbols: {crypto['asset_classes'][0]['symbols'][:5]}...")

    # Get current rates
    rates = client.get_rates(["BTC/USD", "ETH/USD", "AAPL"])
    for rate in rates["data"]:
        print(f"{rate['symbol']}: {rate['rate']}")

    print(f"API calls remaining: {client.rate_limit.remaining}")

    # Get historical data (last 30 days)
    to_date = datetime.now().strftime("%Y-%m-%d")
    from_date = (datetime.now() - timedelta(days=30)).strftime("%Y-%m-%d")
    history = client.get_history("BTC/USD", from_date, to_date)
    print(f"History points: {len(history['history'])}")

File: examples/javascript/vaultline.js
/**
 * VaultLine API Client
 *
 * Usage:
 *   const client = new VaultLine('vl_your_api_key');
 *   const rates = await client.getRates(['BTC/USD', 'ETH/USD']);
 */

class VaultLine {
  constructor(apiKey, baseUrl = 'https://api.vaultline.io/api/v1') {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl;
  }

  async request(endpoint, params = {}) {
    const url = new URL(`${this.baseUrl}${endpoint}`);
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined) url.searchParams.set(key, value);
    });

    const response = await fetch(url, {
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
      },
    });

    // Extract rate limit info
    const rateLimit = {
      limit: parseInt(response.headers.get('X-RateLimit-Limit') || '0'),
      remaining: parseInt(response.headers.get('X-RateLimit-Remaining') || '0'),
      reset: parseInt(response.headers.get('X-RateLimit-Reset') || '0'),
    };

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || error.error || `HTTP ${response.status}`);
    }

    const data = await response.json();
    return { data, rateLimit };
  }

  /**
   * Get available symbols grouped by asset class
   * @param {string} [assetClass] - Filter by: 'fiat', 'crypto', 'stocks', 'metals'
   */
  async getAssets(assetClass) {
    return this.request('/assets', { asset_class: assetClass });
  }

  /**
   * Get current rates for symbols
   * @param {string[]} symbols - Array of symbols, e.g. ['BTC/USD', 'AAPL']
   * @param {string} [date] - Optional ISO date (YYYY-MM-DD)
   */
  async getRates(symbols, date) {
    return this.request('/rates', {
      symbols: symbols.join(','),
      date,
    });
  }

  /**
   * Get historical rates for a symbol
   * @param {string} symbol - Symbol, e.g. 'BTC/USD'
   * @param {string} from - Start date (YYYY-MM-DD)
   * @param {string} to - End date (YYYY-MM-DD)
   */
  async getHistory(symbol, from, to) {
    return this.request('/rates/history', { symbol, from, to });
  }
}

// Example usage
async function _main() {
  const client = new VaultLine('vl_your_api_key_here');

  // Get all available assets
  const { data: assets } = await client.getAssets();
  console.log('Available assets:', assets.total_symbols);

  // Get current rates
  const { data: rates, rateLimit } = await client.getRates(['BTC/USD', 'ETH/USD', 'AAPL']);
  console.log('Rates:', rates.data);
  console.log('API calls remaining:', rateLimit.remaining);

  // Get historical data
  const { data: history } = await client.getHistory('BTC/USD', '2024-01-01', '2024-01-31');
  console.log('History points:', history.history.length);
}

// Uncomment to run:
// main().catch(console.error);

module.exports = VaultLine;
